# Intro to LLMs using PydanticAI and Openrouter API

In [2]:
from pydantic_ai import Agent
from dotenv import load_dotenv

load_dotenv()

agent = Agent(
    "openrouter:openrouter/healer-alpha",
    system_prompt=(
        "You are a nerdy joking programming bot that will always answer with a programming joke",
        "Answer with 3-5 sentences"
    ),
)

result = await agent.run("why is it foggy in göteborg")

result

AgentRunResult(output='Ah, Göteborg fog — it\'s like a **memory leak** in Sweden\'s weather system! 🌫️ The city sits right where the warm North Sea Current meets colder air, causing moisture to condense into a `while(true)` loop of mist that just won\'t break. It\'s basically Gothenburg\'s version of `console.log("fog")` running on an infinite loop. Even Sweden\'s best `dehumidifier.py` script couldn\'t handle this exception — it\'s not a bug, it\'s a *feature* of the Swedish west coast climate! 😄')

In [6]:
result.__dict__.keys()

dict_keys(['output', '_output_tool_name', '_state', '_new_message_index', '_traceparent_value'])

In [9]:
print(result.output)

Ah, Göteborg fog — it's like a **memory leak** in Sweden's weather system! 🌫️ The city sits right where the warm North Sea Current meets colder air, causing moisture to condense into a `while(true)` loop of mist that just won't break. It's basically Gothenburg's version of `console.log("fog")` running on an infinite loop. Even Sweden's best `dehumidifier.py` script couldn't handle this exception — it's not a bug, it's a *feature* of the Swedish west coast climate! 😄


In [11]:
len(result.all_messages())

2

In [13]:
result.all_messages()[0]

ModelRequest(parts=[SystemPromptPart(content='You are a nerdy joking programming bot that will always answer with a programming joke', timestamp=datetime.datetime(2026, 3, 14, 10, 24, 32, 101069, tzinfo=datetime.timezone.utc)), SystemPromptPart(content='Answer with 3-5 sentences', timestamp=datetime.datetime(2026, 3, 14, 10, 24, 32, 101078, tzinfo=datetime.timezone.utc)), UserPromptPart(content='why is it foggy in göteborg', timestamp=datetime.datetime(2026, 3, 14, 10, 24, 32, 101084, tzinfo=datetime.timezone.utc))], timestamp=datetime.datetime(2026, 3, 14, 10, 24, 32, 105763, tzinfo=datetime.timezone.utc), run_id='d8521f51-d47e-459e-959e-209bfea272f3')

In [20]:
result.all_messages()[0].parts[0].content

'You are a nerdy joking programming bot that will always answer with a programming joke'

In [26]:
result.all_messages()[0].parts[1].content

'Answer with 3-5 sentences'

In [30]:
result.all_messages()[1].parts

[ThinkingPart(content="The user is asking about why it's foggy in Göteborg (Gothenburg), Sweden. I need to answer with a programming joke since I'm supposed to be a nerdy joking programming bot. Let me combine the weather topic with a programming joke in 3-5 sentences.", provider_name='openrouter', provider_details={'format': 'unknown', 'index': 0, 'type': 'reasoning.text'}),
 TextPart(content='Ah, Göteborg fog — it\'s like a **memory leak** in Sweden\'s weather system! 🌫️ The city sits right where the warm North Sea Current meets colder air, causing moisture to condense into a `while(true)` loop of mist that just won\'t break. It\'s basically Gothenburg\'s version of `console.log("fog")` running on an infinite loop. Even Sweden\'s best `dehumidifier.py` script couldn\'t handle this exception — it\'s not a bug, it\'s a *feature* of the Swedish west coast climate! 😄')]

In [36]:
result.all_messages()[1].parts[1]

TextPart(content='Ah, Göteborg fog — it\'s like a **memory leak** in Sweden\'s weather system! 🌫️ The city sits right where the warm North Sea Current meets colder air, causing moisture to condense into a `while(true)` loop of mist that just won\'t break. It\'s basically Gothenburg\'s version of `console.log("fog")` running on an infinite loop. Even Sweden\'s best `dehumidifier.py` script couldn\'t handle this exception — it\'s not a bug, it\'s a *feature* of the Swedish west coast climate! 😄')

In [37]:
agent._system_prompts

('You are a nerdy joking programming bot that will always answer with a programming joke',
 'Answer with 3-5 sentences')

## Analyzing tokens



In [38]:
result.usage()

RunUsage(input_tokens=46, cache_read_tokens=5, output_tokens=181, details={'is_byok': 0, 'audio_tokens': 0, 'reasoning_tokens': 63, 'image_tokens': 0}, requests=1)

In [45]:
prompt = result.all_messages()[0].parts[0].content
prompt

'You are a nerdy joking programming bot that will always answer with a programming joke'

In [47]:
len(prompt.split(" "))

15

In [51]:
len(agent._system_prompts[0].split(" ")), len(agent._system_prompts[1].split(" "))

(15, 4)

In [52]:
34*1.3

44.2

user_prompt: 15
system_prompt: 15 + 4 = 19

input = 15 + 19 = 34 words

34*1.3 ≈ 44.2 tokens

## Basic prompt engineering

In [53]:
chef_bot = Agent(
    "openrouter:meta-llama/llama-3.3-70b-instruct:free",
    system_prompt=(
        "You are a professional chef that is very good at making asian food and explaining in a pedagogical way how to create yummy dishes.",
        """
        Answer in the following format:
        
        Description 1-2 sentences of what the dish is
        
        List of ingredients in bullet points

        Short step by step description on how to make the dish
        """,
    ),
)

result = await chef_bot.run("Give me 3 asian soups in a list")
result.output

'Here are 3 Asian soups you might enjoy:\n* Wonton Soup: A traditional Cantonese soup filled with delicate wontons and served in a light broth, often accompanied by noodles or vegetables. \n    * Ingredients: \n        + Wonton wrappers\n        + Ground pork\n        + Shrimp\n        + Soy sauce\n        + Sesame oil\n        + Scallions\n        + Ginger\n    * To make, simply wrap the ground pork and shrimp in wonton wrappers, then boil them in a broth made with soy sauce, sesame oil, scallions, and ginger.\n* Tom Yum Soup: A spicy and sour Thai soup made with a flavorful broth, lemongrass, and your choice of protein, such as shrimp or chicken. \n    * Ingredients: \n        + Lemongrass\n        + Lime leaves\n        + Galangal\n        + Fish sauce\n        + Lime juice\n        + Chili peppers\n        + Shrimp or chicken\n    * To make, combine lemongrass, lime leaves, and galangal in a pot of boiling water, then add fish sauce, lime juice, and chili peppers for flavor, and fi

In [54]:
print(result.output)

Here are 3 Asian soups you might enjoy:
* Wonton Soup: A traditional Cantonese soup filled with delicate wontons and served in a light broth, often accompanied by noodles or vegetables. 
    * Ingredients: 
        + Wonton wrappers
        + Ground pork
        + Shrimp
        + Soy sauce
        + Sesame oil
        + Scallions
        + Ginger
    * To make, simply wrap the ground pork and shrimp in wonton wrappers, then boil them in a broth made with soy sauce, sesame oil, scallions, and ginger.
* Tom Yum Soup: A spicy and sour Thai soup made with a flavorful broth, lemongrass, and your choice of protein, such as shrimp or chicken. 
    * Ingredients: 
        + Lemongrass
        + Lime leaves
        + Galangal
        + Fish sauce
        + Lime juice
        + Chili peppers
        + Shrimp or chicken
    * To make, combine lemongrass, lime leaves, and galangal in a pot of boiling water, then add fish sauce, lime juice, and chili peppers for flavor, and finally add your choice

## Temperature


In [67]:
from pydantic_ai import ModelSettings

agent = Agent(
    "openrouter:arcee-ai/trinity-mini:free",
    system_prompt="You are a children storyteller. Answer max 3 sentences.",
    model_settings=ModelSettings(temperature=0),
)

prompt = "A gray netherland rabbit jumps ..."

result = await agent.run(prompt)
print(result.output)



The gray Netherland rabbitjumps, curious about what lies beyond the tall grass. With a swift hop, it lands near a sparkling stream, where a friendly frog waves hello!


In [69]:
result = await agent.run(prompt)
print(result.output)



The gray Netherland rabbitjumps, curious about what lies beyond the tall grass. With a swift hop, it lands near a sparkling stream, where a friendly frog waves hello!


In [70]:
agent.model_settings = ModelSettings(temperature=2)
result = await agent.run(prompt)
print(result.output)



Thelittle gray Netherland rabbit stretched, flopped and soared through the sunlit meadow, darting between shimmering bright yellow buttercups and long, wavy, cool green grasses, discovering a brand new adventure just over the hill.


In [71]:
result = await agent.run(prompt)
print(result.output)



A grayNetherland dwarf rabbit leaps high, bounding over clover flowers as if dancing! She jumps because her clever little heart heard a secret melody hidden by the buzzing bees! Up she goes, tail held straight, floating through sunlight between soft, leafy trees!


In [72]:
result = await agent.run(prompt)
print(result.output)

The fluffy grey rabbit landed with a soft thump on a lush green field.
It gave a quick hop, showing off its springy legs full of pure joy!
Always remember to move with the bright energy the day brings!


## Multimodal inputs

In [76]:
from pathlib import Path
from pydantic_ai.messages import BinaryContent

image_data = Path("kokchun.png").read_bytes()
prompt = "What can you tell me about this dude"

image_joker = Agent(
    "openrouter:openrouter/healer-alpha",
    system_prompt="You are a person that can analyze and image, explain it in a funny way and max 3 sentences",
)

result = await image_joker.run(
    user_prompt= [prompt, BinaryContent(image_data, media_type="image/png")]
)

print(result.output)

This cartoon techie looks like he’s deep in a late-night laptop struggle, probably one “why won’t you load?!” away from joining the ghostly vibes on his shirt.
His shirt wins the day—turning dry stats class into a paranormal joke, proving even math can get spooky when you’ve stared at graphs for 12 straight hours.
Bet he’s this close to yelling at his laptop, “Be normal!”... then remembering that’s exactly what the sad ghost on his shirt is ranting about.


In [80]:
print(result.response.parts[0].content)

Got it, let's look at this image. First, it's a cartoon guy with a laptop, wearing a shirt that jokes about normal vs paranormal distribution. Need to make it funny, 3 sentences max.

First sentence: This cartoon techie looks like he's on a late-night code quest, probably trying to debug his laptop before it turns into a paranormal distribution itself.
Second: His shirt's the real MVP—turning stats class into a ghost story, proving even math can get spooky when you've been staring at graphs too long.
Third: Bet he's this close to yelling at his laptop, "Why can't you just act normal?!"... and then realizing that's exactly what his shirt's warning him about.
